src/search/merge_and_shrink/labels.cc

```diff
@@ -12,50 +12,64 @@
 using namespace std;
 
 namespace merge_and_shrink {
-Labels::Labels(vector<unique_ptr<Label>> &&labels)
-    : labels(move(labels)),
-      max_size(0) {
-    if (!this->labels.empty()) {
-        max_size = this->labels.size() * 2 - 1;
+LabelsConstIterator::LabelsConstIterator(
+    const vector<int> &label_costs, bool end)
+    : label_costs(label_costs),
+      current_index(end ? label_costs.size() : 0) {
+    next_valid_index();
+}
+
+void LabelsConstIterator::next_valid_index() {
+    while (current_index < label_costs.size()
+           && label_costs[current_index] == -1) {
+        ++current_index;
     }
 }
 
-void Labels::reduce_labels(const vector<int> &old_label_nos) {
+void LabelsConstIterator::operator++() {
+    ++current_index;
+    next_valid_index();
```

Comment:
Change this to `advance_to_next_valid_index()` or similar? This sounds more like a method that would return the next valid index.